### PR TITLE
cassandra 4.0.0

### PR DESCRIPTION
**Pull request by branchv**
Created with `brew bump-formula-pr`.

~`resource` blocks may require updates.~

### CODE REVIEW

File: Formula/cassandra.rb [hunk 1]
```diff
@@ -13,25 +13,17 @@ class Cassandra < Formula
...
   depends_on "cython" => :build
-  # Due to Python 2 (https://issues.apache.org/jira/browse/CASSANDRA-10190), cassandra 4 will support python3
-  depends_on :macos
-  depends_on "openjdk@8" # cassandra 4 will support Java 11
+  depends_on "python@3.9" => :build
+  depends_on "openjdk@11"
 
-  # Only >=Yosemite has new enough setuptools for successful compile of the below deps.
-  # Python 2 needs setuptools < 45.0.0 (https://github.com/pypa/setuptools/issues/2094)
   resource "setuptools" do
```
Comment (carlocab) [inline Formula/cassandra.rb:19]: We shouldn't need this anymore -- Python `venv`s have `setuptools` installed by default.

File: Formula/cassandra.rb [hunk 2]
```diff
@@ -13,25 +13,17 @@ class Cassandra < Formula
...
+    url "https://files.pythonhosted.org/packages/db/e2/c0ced9ccffb61432305665c22842ea120c0f649eec47ecf2a45c596707c4/setuptools-57.4.0.tar.gz"
+    sha256 "6bac238ffdf24e8806c61440e755192470352850f3419a52f26ffe0a1a64f465"
   end
 
   resource "six" do
-    url "https://files.pythonhosted.org/packages/6b/34/415834bfdafca3c5f451532e8a8d9ba89a21c9743a0c59fbd0205c7f9426/six-1.15.0.tar.gz"
-    sha256 "30639c035cdb23534cd4aa2dd52c3bf48f06e5f4a941509c8bafd8ce11080259"
+    url "https://files.pythonhosted.org/packages/71/39/171f1c67cd00715f190ba0b100d606d440a28c93c7714febeca8b79af85e/six-1.16.0.tar.gz"
+    sha256 "1e61c37477a1626458e36f7b1d82aa5c9b094fa4802892072e49de9c60c4c926"
   end
```
Comment (carlocab) [inline Formula/cassandra.rb:27]: Let's do `depends_on "six"` instead.

File: Formula/cassandra.rb [hunk 3]
```diff
@@ -46,19 +38,20 @@ class Cassandra < Formula
...
-    pypath = libexec/"vendor/lib/python2.7/site-packages"
+    xy = Language::Python.major_minor_version "python3"
+    pypath = libexec/"vendor/lib/python#{xy}/site-packages"
     ENV.prepend_create_path "PYTHONPATH", pypath
     resources.each do |r|
       r.stage do
-        system "python", *Language::Python.setup_install_args(libexec/"vendor")
+        system "python3", *Language::Python.setup_install_args(libexec/"vendor")
       end
     end
```
Comment (carlocab) [inline Formula/cassandra.rb:56]: We can probably do something like this here:
```ruby
venv = virtualenv_create(libexec/"vendor", "python3")
pypath = libexec/"vendor"/Language::Python.site_packages("python3")
venv.pip_install resources
```
We can also probably move the definition of `pypath` closer to where it is actually used.

File: Formula/cassandra.rb (hunk 2 above)
Comment (branchvincent) [reply Formula/cassandra.rb:27]: thanks for the help! i may have missed something, but it doesn't seem like `six` is installed into the venv this way?

Comment (carlocab) [reply Formula/cassandra.rb]: You're right that it's not, but `venv`s that `brew` creates [have access to system site-packages](https://github.com/Homebrew/brew/blob/3e52e8b48ab036af77a582219ea250107c84432f/Library/Homebrew/language/python.rb#L221).

File: Formula/cassandra.rb [hunk 4]
```diff
@@ -136,6 +116,7 @@ def install
     bin.write_exec_script Dir["#{libexec}/bin/*"]
 
     rm %W[#{bin}/cqlsh #{bin}/cqlsh.py] # Remove existing exec scripts
+    pypath = libexec/"vendor"/Language::Python.site_packages("python3")
     (bin/"cqlsh").write_env_script libexec/"bin/cqlsh", PYTHONPATH: pypath
     (bin/"cqlsh.py").write_env_script libexec/"bin/cqlsh.py", PYTHONPATH: pypath
```
Comment (carlocab) [inline Formula/cassandra.rb:121]: I _think_ this should fix it. Also, the `rm` call is probably redundant.
```suggestion
    venv_python = libexec/"vendor/bin/python"
    (bin/"cqlsh").write_env_script libexec/"bin/cqlsh", CQLSH_PYTHON: venv_python
    rewrite_shebang python_shebang_rewrite_info(venv_python), libexec/"bin/cqlsh.py"
```
We need to make sure that the Python interpreter being used is the one in our `venv`. We'll need to do
```ruby
include Language::Python::Shebang
```
too.

Comment (carlocab) [reply Formula/cassandra.rb]: Nope, `CQLSH_PYTHON` doesn't behave the way I expect it to. Upon inspection, it seems to be discarded immediately after checking its existence, which is strange. Tried something else.

Comment (carlocab) [reply Formula/cassandra.rb]: Hmmm... Haven't managed to persuade this to use the `venv` python. Weird.

**Comment by carlocab** [comment]
Not sure why my changes aren't working on CI. They work for me locally.

**Comment by carlocab** [comment]
As an aside, I think `cassandra-driver` needs these extra resources:
```ruby
  resource "cassandra-driver" do
    url "https://files.pythonhosted.org/packages/af/aa/3d3a6dae349d4f9b69d37e6f3f8b8ef286a06005aa312f0a3dc7af0eb556/cassandra-driver-3.25.0.tar.gz"
    sha256 "8ad7d7c090eb1cac6110b3bfc1fd2d334ac62f415aac09350ebb8d241b7aa7ee"
  end

  resource "click" do
    url "https://files.pythonhosted.org/packages/21/83/308a74ca1104fe1e3197d31693a7a2db67c2d4e668f20f43a2fca491f9f7/click-8.0.1.tar.gz"
    sha256 "8c04c11192119b1ef78ea049e0a6f0463e4c48ef00a30160c704337586f3ad7a"
  end

  resource "geomet" do
    url "https://files.pythonhosted.org/packages/cf/21/58251b3de99e0b5ba649ff511f7f9e8399c3059dd52a643774106e929afa/geomet-0.2.1.post1.tar.gz"
    sha256 "91d754f7c298cbfcabd3befdb69c641c27fe75e808b27aa55028605761d17e95"
  end
```

We can also update `thrift`:
```ruby
  resource "thrift" do
    url "https://files.pythonhosted.org/packages/97/1e/3284d19d7be99305eda145b8aa46b0c33244e4a496ec66440dac19f8274d/thrift-0.13.0.tar.gz"
    sha256 "9af1c86bf73433afc6010ed376a6c6aca2b54099cc0d61895f640870a9ae7d89"
  end
```

**Comment by carlocab** [comment]
Ok, I see the problem now. This should fix it.

**Review by SMillerDev — APPROVED** [review]

File: Formula/cassandra.rb (hunk 2 above)
Comment (branchvincent) [reply Formula/cassandra.rb]: ah i see now

**Comment by branchvincent** [comment]
@carlocab thanks a lot for helping me here!!

**Pull request closed without merging**